server/src/main/java/io/spine/server/event/enrich/MessageEnrichment.java

```diff
@@ -104,74 +81,111 @@ public T apply(S eventMsg, C context) {
         checkNotNull(eventMsg);
         checkNotNull(context);
         ensureActive();
-        verifyOwnState();
+        verifyState();
 
-        T defaultTarget = defaultInstance(getEnrichmentClass());
-        Message.Builder builder = defaultTarget.toBuilder();
-        setFields(builder, eventMsg, context);
-        @SuppressWarnings("unchecked") // types are checked during the initialization and validation
-        T result = (T) builder.build();
+        Helper<S, C, T> helper = new Helper<>(this, eventMsg, context);
+        T result = helper.createEnrichment();
         return result;
     }
 
     private void markActive() {
         try {
-            verifyOwnState();
+            verifyState();
             active = true;
         } catch (RuntimeException ignored) {
             active = false;
         }
     }
 
-    private void verifyOwnState() {
+    private void verifyState() {
         checkNotNull(fieldMap, "fieldMap");
         checkNotNull(fieldFunctions, "fieldFunctions");
         checkState(!fieldMap.isEmpty(), "fieldMap is empty");
         checkState(!fieldFunctions.isEmpty(), "fieldFunctions is empty");
     }
 
-    @SuppressWarnings({
-            "ConstantConditions" /* it is assured that collections are not null, and
-                                    after validation maps have required entries. */,
-            "MethodWithMultipleLoops"}
-    )
-    private void setFields(Message.Builder builder, S sourceMessage, C context) {
-        for (FieldDescriptor srcField : fieldMap.keySet()) {
-            Object srcFieldValue = getSrcFieldValue(srcField, sourceMessage, context);
-            Class<?> sourceFieldClass = srcFieldValue.getClass();
-            Collection<EnrichmentFunction<?, ?, ?>> functions =
-                    fieldFunctions.get(sourceFieldClass);
-            Collection<FieldDescriptor> targetFields = fieldMap.get(srcField);
+    /**
+     * Method object for {@link MessageEnrichment#apply(Message, Message)}.
+     *
+     * @param <S> the type of the source message
+     * @param <T> the type of the target message
+     * @param <C> the type of the source message context
+     */
+    private static final class Helper<S extends Message, C extends Message, T extends Message> {
```

Comment:
A message context should always implement `MessageContext` interface. Please use it instead of plain `Message`.